handler/graphql.go

```diff
@@ -390,187 +388,199 @@ func computeQueryHash(query string) string {
 	return hex.EncodeToString(b[:])
 }
 
-func (gh *graphqlHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	if r.Method == http.MethodOptions {
-		w.Header().Set("Allow", "OPTIONS, GET, POST")
-		w.WriteHeader(http.StatusOK)
-		return
-	}
-
-	if strings.Contains(r.Header.Get("Upgrade"), "websocket") {
-		connectWs(gh.exec, w, r, gh.cfg, gh.cache)
-		return
-	}
-
-	w.Header().Set("Content-Type", "application/json")
-	var reqParams params
-	switch r.Method {
-	case http.MethodGet:
-		reqParams.Query = r.URL.Query().Get("query")
-		reqParams.OperationName = r.URL.Query().Get("operationName")
+func (gh *graphqlHandler) serve(reqParams graphql.ParsedParams) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 
-		if variables := r.URL.Query().Get("variables"); variables != "" {
-			if err := jsonDecode(strings.NewReader(variables), &reqParams.Variables); err != nil {
-				sendErrorf(w, http.StatusBadRequest, "variables could not be decoded")
-				return
-			}
+		if strings.Contains(r.Header.Get("Upgrade"), "websocket") {
+			connectWs(gh.exec, w, r, gh.cfg, gh.cache)
+			return
 		}
 
-		if extensions := r.URL.Query().Get("extensions"); extensions != "" {
-			if err := jsonDecode(strings.NewReader(extensions), &reqParams.Extensions); err != nil {
-				sendErrorf(w, http.StatusBadRequest, "extensions could not be decoded")
+		ctx := r.Context()
+
+		var queryHash string
+		apqRegister := false
+		apq := reqParams.Extensions != nil && reqParams.Extensions.PersistedQuery != nil
+		if apq {
```

Comment:
Will be easier to see what's changed in here once this flattens back out to the same level, this could be a middleware stack.

imagine
```go
handler.Use(func(next graphql.HandlerFunc)) {
   return graphql.HandlerFunc(func(w, r, params)) {
        /// apq stuff that modifies params
        next(w, r, modifiedParams)
   }
}
```
not inline of course, APQ would just be a function that could be used.

This is really the interface we need to get right for any of this to go ahead IMO.